bin/build-calypso-app.mjs

```diff
@@ -0,0 +1,121 @@
+/* eslint-disable import/no-nodejs-modules */
+/* eslint-disable no-console */
+/* eslint-disable no-process-exit */
+import { exec } from 'child_process';
+import chokidar from 'chokidar';
+import runAll from 'npm-run-all';
+import yargs from 'yargs';
+import { hideBin } from 'yargs/helpers';
+
+// Default NODE_ENV is development unless manually set to production.
+if ( ! process.env.NODE_ENV ) {
+	process.env.NODE_ENV = 'development';
+}
+if ( ! process.env.BROWSERSLIST_ENV ) {
+	process.env.BROWSERSLIST_ENV = 'wpcom';
+}
+
+const { argv } = yargs( hideBin( process.argv ) ).options( {
+	watch: { type: 'boolean', default: false },
+	sync: { type: 'boolean', default: false, implies: 'remotePath' },
+	remotePath: { type: 'string' },
+	localPath: {
+		type: 'string',
+		required: true,
+		coerce: ( distPath ) => ( distPath ? `${ process.cwd() }/${ distPath }/` : undefined ),
+	},
+	verbose: { type: 'boolean', default: false, alias: 'v' },
+} );
+await runBuilder( argv );
+
+async function runBuilder( { watch, sync, localPath, remotePath, verbose } ) {
+	if ( verbose ) {
+		console.info( `Watch mode: ${ watch ? 'yes' : 'no' }\nSync mode: ${
+			sync ? 'yes' : 'no'
+		}\nLocal path: ${ localPath }\nRemote path: ${ remotePath }
+		` );
+	}
+
+	const runOpts = {
+		stdout: process.stdout,
+		stderr: process.stderr,
+		printLabel: true,
+	};
+	console.log( 'Cleaning...' );
+	await runAll( 'clean', runOpts );
+
+	console.log( 'Starting webpack...' );
+	// Run build tasks in parallel
+	runAll( [ 'build:*' ], { ...runOpts, parallel: true } )
+		.then( () => {
+			console.log( 'Build completed!' );
+			if ( ! watch && sync ) {
+				// In non-watch + sync mode, we sync only once after the build has finished.
+				syncToRemote( localPath, remotePath );
+			}
+		} )
+		.catch( ( e ) => {
+			console.error( 'The build failed.' );
+			console.error( `Reported build error: ${ e.message }` );
+			showTips( e.results );
+			process.exit( 1 );
+		} );
+
+	// In dev mode, we start watching to sync while the webpack build is happening.
+	if ( watch && sync ) {
+		syncToRemote( localPath, remotePath, true );
+	}
+}
+
+function syncToRemote( localPath, remotePath, watch = false ) {
+	const debouncedSync = debouncer( () => {
+		console.info( 'Performing sync...' );
+		exec( `rsync -ahz --exclude=".*" ${ localPath } wpcom-sandbox:${ remotePath }` );
```

Comment:
`rsync` uses SSH to connect to the remote host, and uses the host aliases defined in my SSH config, like `wpcom-sandbox`? And the `wpcom-sandbox` name is hardcoded? It should be overridable with an env variable.

About the debouncing logic:
- is it safe to start a new `rsync` even if the previous one hasn't finished yet? Can happen if it takes more than 500ms
- should we really start the `rsync` immediately on first call, without waiting? Then it will start right when webpack emits the very first file, and will sync while webpack is still writing. A better logic would be: after a first write, schedule a sync, and perform it only after writes have ceased for at least 500ms. That's exactly what the plain Lodash `_.debounce( sync, 500 )` would do.

Reply:
> is it safe to start a new rsync even if the previous one hasn't finished yet? Can happen if it takes more than 500ms

I've reworked the logic to kill the existing rsync process before trying a new sync. I think this should be ok

> should we really start the rsync immediately on first call, without waiting

Hmm, yeah, we don't have to do an initial sync. What if `chokidar` didn't register any change events after the sync logic got set up? This could theoretically happen if the webpack build finishes before the call to `chokidar`. That does seem pretty unlikely, though.